lib/tools/te_string.h

```diff
@@ -810,6 +811,7 @@ static inline bool
 te_substring_is_valid(const te_substring_t *substr)
 {
     return substr != NULL && substr->base != NULL &&
+        substr->base->ptr != NULL &&
```

Comment:
That's utterly not correct, because a substring that points beyond the end of string should always be considered correct, otherwise appending to an empty string would be broken. The definition of `te_substing_find`, on the contrary, _does not_, indeed, cover this border case, and many similiar functions as well.